hs_composite_resource/models.py

```diff
@@ -127,6 +212,145 @@ def can_set_folder_to_fileset(self, dir_path):
         # if there are any files in the dir_path, we can set the folder to fileset aggregation
         return len(files_in_path) > 0
 
+    def can_set_folder_to_model_instance_aggregation(self, dir_path):
+        """Checks if the specified folder *dir_path* can be set to ModelInstance aggregation
+
+        :param dir_path: Resource file directory path (full folder path starting with resource id)
+        for which the ModelInstance aggregation to be set
+
+        :return
+        If the specified folder is already represents an aggregation, return Flase
+        if the specified folder path does not contain any files, return False
+        if any of the files in the specified folder path is part of an aggregation other than fileset, return False
+        if any of the sub-folders is a fileset aggregation, returns False
+        if any of the parent folders is a model program aggregation, return False
+        if any of the parent folders is a model instance aggregation, return False
+        otherwise, return True
+        """
+        return self._can_set_folder_to_mi_or_mp_aggregation(dir_path=dir_path, aggr_type="ModelInstance")
+
+    def can_set_folder_to_model_program_aggregation(self, dir_path):
+        """Checks if the specified folder *dir_path* can be set to ModelProgram aggregation
+
+        :param dir_path: Resource file directory path (full folder path starting with resource id)
+        for which the ModelProgram aggregation to be set
+
+        :return
+        If the specified folder is already represents an aggregation, return False
+        if the specified folder does not contain any files, return False
+        if any of the files in the specified folder path is part of an aggregation other than fileset, return False
+        if any of the sub-folders is a fileset aggregation, return False
+        if any of the sub-folders is a model program aggregation, return False
+        if any of the sub-folders is a model instance aggregation, return False
+        if any of the parent folders is a model program aggregation, return False
+        otherwise, return True
+        """
+        return self._can_set_folder_to_mi_or_mp_aggregation(dir_path=dir_path, aggr_type="ModelProgram")
+
+    def _can_set_folder_to_mi_or_mp_aggregation(self, dir_path, aggr_type):
```

Comment:
The logic in this method needs to be reorganized.  There is a whole lot going on with lots of exceptions to the logical flow.